src/workers/tpu_utils/tpu_connection_manager.rs

```diff
@@ -116,50 +121,162 @@ impl ActiveConnection {
         exit_signal: Arc<AtomicBool>,
-    ) -> Option<SendStream> {
-        loop {
-            if let Some(connection) = connection {
-                match timeout(
-                    QUIC_CONNECTION_TIMEOUT_DURATION_IN_SEC,
-                    connection.open_uni(),
-                )
-                .await
-                {
-                    Ok(Ok(unistream)) => return Some(unistream),
-                    Ok(Err(_)) => (),
-                    Err(_) => return None,
+        last_stable_id: Arc<AtomicU64>,
+    ) {
+        for _ in 0..3 {
+            let conn = {
+                let last_stable_id = last_stable_id.load(Ordering::Relaxed) as usize;
+                let conn = connection.read().await;
+                if conn.stable_id() == last_stable_id {
+                    // problematic connection
+                    drop(conn);
+                    let mut conn = connection.write().await;
+                    let new_conn = Self::connect(
+                        identity,
+                        true,
+                        endpoint.clone(),
+                        socket_addr.clone(),
+                        exit_signal.clone(),
+                    )
+                    .await;
+                    if let Some(new_conn) = new_conn {
+                        *conn = new_conn;
+                        conn.clone()
+                    } else {
+                        // could not connect
+                        return;
+                    }
+                } else {
+                    conn.clone()
                 }
-            } else {
-                reconnect = true
-            }
+            };
+            let mut retry = false;
+            for tx in &txs {
+                let (stream, retry_conn) =
+                    Self::open_unistream(conn.clone(), last_stable_id.clone()).await;
+                if let Some(send_stream) = stream {
+                    if exit_signal.load(Ordering::Relaxed) {
+                        return;
+                    }
 
-            if !reconnect {
-                return None;
+                    retry = Self::write_all(
+                        send_stream,
+                        tx,
+                        identity,
+                        last_stable_id.clone(),
+                        conn.stable_id() as u64,
+                    )
+                    .await;
+                } else {
+                    retry = retry_conn;
+                }
+            }
+            if !retry {
+                break;
             }
+        }
+    }
 
-            // re connect
-            let Some(conn) = Self::connect(
-                identity,
-                already_connected,
-                endpoint.clone(),
-                addr,
-                exit_signal.clone(),
-            )
-            .await else {
-                return None;
-            };
+    // copied from solana code base
+    fn compute_receive_window_ratio_for_staked_node(
```

Comment:
Is this function not exported from any crate from solana ?

Reply:
It is not public but private function.